hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/StakingPeriodTimeHook.java

```diff
@@ -0,0 +1,104 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.workflows;
+
+import static com.hedera.node.app.service.mono.ledger.accounts.staking.StakePeriodManager.DEFAULT_STAKING_PERIOD_MINS;
+import static com.hedera.node.app.service.mono.utils.Units.MINUTES_TO_MILLISECONDS;
+import static com.swirlds.common.stream.LinkedObjectStreamUtilities.getPeriod;
+import static java.time.ZoneOffset.UTC;
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hedera.node.app.service.token.impl.handlers.staking.EndOfStakingPeriodUpdater;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.config.data.StakingConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Instant;
+import java.time.LocalDate;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A {@link ConsensusTimeHook} implementation that handles the daily staking period updates
+ */
+@Singleton
+public class StakingPeriodTimeHook implements ConsensusTimeHook {
+    private static final Logger logger = LogManager.getLogger(StakingPeriodTimeHook.class);
+
+    private final EndOfStakingPeriodUpdater stakingCalculator;
+    private Instant consensusTimeOfLastHandledTxn;
+
+    @Inject
+    public StakingPeriodTimeHook(@NonNull final EndOfStakingPeriodUpdater stakingPeriodCalculator) {
+        this.stakingCalculator = requireNonNull(stakingPeriodCalculator);
+    }
+
+    /**
+     * This time hook is responsible for performing necessary staking updates and distributing staking
+     * rewards. This should only be done during handling of the first transaction of each new staking
+     * period, which staking period usually starts at midnight UTC.
+     *
+     * <p>The only exception to this rule is when {@code consensusTimeOfLastHandledTxn} is null,
+     * <b>which should only happen on node startup.</b> The node should therefore run this process
+     * to catch up on updates and distributions when first coming online.
+     */
+    @Override
+    public void process(@NonNull Instant consensusTime, @NonNull final HandleContext context) {
+        if (consensusTimeOfLastHandledTxn == null
+                || consensusTime.getEpochSecond() > consensusTimeOfLastHandledTxn.getEpochSecond()
+                        && isNextStakingPeriod(consensusTime, consensusTimeOfLastHandledTxn, context)) {
+            // Handle the daily staking distributions and updates
+            try {
+                stakingCalculator.updateNodes(consensusTime, context);
```

Comment:
Every midnight with the same condition 
```
if (consensusTimeOfLastHandledTxn == null
                || consensusTime.getEpochSecond() > consensusTimeOfLastHandledTxn.getEpochSecond()
                        && isNextStakingPeriod(consensusTime, consensusTimeOfLastHandledTxn, context))
```
we should also update `ExchangeRate` here. So this file name can probably be renamed to something that happens at midnight UTC ?
We can add TODO to update `ExchangeRate` there

Reply:
My understanding was that the exchange rate is updated by uploading file `0.0.112`. Are there multiple ways to update the exchange rate?

Reply:
@netopyr That is correct. I mean about this step, [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/state/logic/NetworkCtxManager.java#L141), where we replace the `MerkleNetworkCtx` midnightRates with new one

Reply:
IMO that work should be separate from this work. Instead of doing everything in a single hook class, the idea is to use `ConsensusTimeHook` – an interface – to implement things like staking or updating the exchange rate, but in separate classes. That way the logic isn't all thrown together in a big mush. Thus, this class would only be for handling staking updates

Reply:
Ah yes, makes sense. I thought this will be re-used.

Reply:
Good not to forget, though. I'll add a ticket and FUTURE note to do this work

Reply:
Added a ticket (#7836) to address, and a comment with the issue number

Reply:
Thanks @Neeharika-Sompalli , I was not aware of this mechanism.